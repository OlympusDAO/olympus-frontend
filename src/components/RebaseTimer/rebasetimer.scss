.rebase-timer {
  position: relative;
  width: max-content;
}
