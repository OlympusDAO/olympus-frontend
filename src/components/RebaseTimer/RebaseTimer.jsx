import { useSelector, useDispatch } from "react-redux";
import { getRebaseBlock, secondsUntilBlock, prettifySeconds } from "../../helpers";
import { Box, Typography } from "@material-ui/core";
import "./rebasetimer.scss";
import { Skeleton } from "@material-ui/lab";
import { useEffect, useMemo, useState } from "react";
import { loadAppDetails } from "../../slices/AppSlice";
import { useWeb3Context } from "../../hooks/web3Context";
import { Trans } from "@lingui/macro";

function RebaseTimer() {
  const dispatch = useDispatch();
  const { provider, chainID } = useWeb3Context();

  const SECONDS_TO_REFRESH = 60;
  const [secondsToRebase, setSecondsToRebase] = useState(0);
  const [rebaseString, setRebaseString] = useState("");
  const [secondsToRefresh, setSecondsToRefresh] = useState(SECONDS_TO_REFRESH);

  const currentBlock = useSelector(state => {
    return state.app.currentBlock;
  });

  function initializeTimer() {
    const rebaseBlock = getRebaseBlock(currentBlock);
    const seconds = secondsUntilBlock(currentBlock, rebaseBlock);
    setSecondsToRebase(seconds);
    const prettified = prettifySeconds(seconds);
    setRebaseString(prettified !== "" ? prettified : <Trans>Less than a minute</Trans>);
  }

  // This initializes secondsToRebase as soon as currentBlock becomes available
  useMemo(() => {
    if (currentBlock) {
      initializeTimer();
    }
  }, [currentBlock]);

  // After every period SECONDS_TO_REFRESH, decrement secondsToRebase by SECONDS_TO_REFRESH,
  // keeping the display up to date without requiring an on chain request to update currentBlock.
  useEffect(() => {
    let interval = null;
    if (secondsToRefresh > 0) {
      interval = setInterval(() => {
        setSecondsToRefresh(secondsToRefresh => secondsToRefresh - 1);
      }, 1000);
    } else {
      // When the countdown goes negative, reload the app details and reinitialize the timer
      if (secondsToRebase < 0) {
        async function reload() {
          await dispatch(loadAppDetails({ networkID: chainID, provider: provider }));
        }
        reload();
        setRebaseString("");
      } else {
        clearInterval(interval);
        setSecondsToRebase(secondsToRebase => secondsToRebase - SECONDS_TO_REFRESH);
        setSecondsToRefresh(SECONDS_TO_REFRESH);
        const prettified = prettifySeconds(secondsToRebase);
        setRebaseString(prettified !== "" ? prettified : <Trans>Less than a minute</Trans>);
      }
    }
    return () => clearInterval(interval);
  }, [secondsToRebase, secondsToRefresh]);

  return (
    <Box className="rebase-timer">
      <Typography variant="body2">
        {currentBlock ? (
          secondsToRebase > 0 ? (
            <>
              <strong>{rebaseString}&nbsp;</strong>
              <Trans>to next rebase</Trans>
            </>
          ) : (
            <strong>rebasing</strong>
          )
        ) : (
          <Skeleton width="155px" />
        )}
      </Typography>
    </Box>
  );
}

export default RebaseTimer;
