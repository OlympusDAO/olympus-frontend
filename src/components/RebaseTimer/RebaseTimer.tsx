import "./RebaseTimer.scss";

import { Trans } from "@lingui/macro";
import { Box, Typography } from "@material-ui/core";
import { Skeleton } from "@material-ui/lab";
import React, { useEffect, useMemo, useState } from "react";
import { useDispatch } from "react-redux";

import { getRebaseBlock, prettifySeconds } from "../../helpers";
import { useAppSelector } from "../../hooks";
import { useWeb3Context } from "../../hooks/web3Context";
import { loadAppDetails } from "../../slices/AppSlice";

const SECONDS_TO_REFRESH = 60;

const RebaseTimer: React.FC = () => {
  const dispatch = useDispatch();
  const { provider, networkId } = useWeb3Context();

  const [secondsToRebase, setSecondsToRebase] = useState<number>(0);
  const [rebaseString, setRebaseString] = useState<string | React.ReactElement>("");
  const [secondsToRefresh, setSecondsToRefresh] = useState<number>(SECONDS_TO_REFRESH);

  const currentBlock = useAppSelector(state => state.app.currentBlock);
  const secondsToEpoch = useAppSelector(state => state.app.secondsToEpoch);

  // This initializes secondsToRebase as soon as currentBlock becomes available
  useMemo(() => {
    if (secondsToEpoch && currentBlock) {
      const rebaseBlock = getRebaseBlock(currentBlock);
      setSecondsToRebase(secondsToEpoch);
      const prettified = prettifySeconds(secondsToEpoch);
      setRebaseString(prettified !== "" ? prettified : <Trans>Less than a minute</Trans>);
    }
  }, [secondsToEpoch]);

  // After every period SECONDS_TO_REFRESH, decrement secondsToRebase by SECONDS_TO_REFRESH,
  // keeping the display up to date without requiring an on chain request to update currentBlock.
  useEffect(() => {
    let interval = 0;
    if (secondsToRefresh > 0) {
      interval = window.setInterval(() => {
        setSecondsToRefresh(secondsToRefresh => secondsToRefresh - 1);
      }, 1000);
    } else {
      // When the countdown goes negative, reload the app details and reinitialize the timer
      if (secondsToRebase < 0) {
        async function reload() {
          await dispatch(loadAppDetails({ networkID: networkId, provider: provider }));
        }
        reload();
        setRebaseString("");
      } else {
        window.clearInterval(interval);
        setSecondsToRebase(secondsToRebase => secondsToRebase - SECONDS_TO_REFRESH);
        setSecondsToRefresh(SECONDS_TO_REFRESH);
        const prettified = prettifySeconds(secondsToRebase);
        setRebaseString(prettified !== "" ? prettified : <Trans>Less than a minute</Trans>);
      }
    }
    return () => window.clearInterval(interval);
  }, [secondsToRebase, secondsToRefresh]);

  return (
    <Box className="rebase-timer">
      <Typography variant="body2">
        {currentBlock ? (
          secondsToRebase > 0 ? (
            <>
              <strong>{rebaseString}&nbsp;</strong>
              <Trans>to next rebase</Trans>
            </>
          ) : (
            <strong>rebasing</strong>
          )
        ) : (
          <Skeleton width="155px" />
        )}
      </Typography>
    </Box>
  );
};

export default RebaseTimer;
