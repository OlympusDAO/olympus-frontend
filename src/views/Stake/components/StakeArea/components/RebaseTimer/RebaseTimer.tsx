import "./RebaseTimer.scss";

import { Trans } from "@lingui/macro";
import { Box, Skeleton, Typography } from "@mui/material";
import React from "react";
import { useNextRebaseDate } from "src/views/Stake/components/StakeArea/components/RebaseTimer/hooks/useNextRebaseDate";

import { prettifySeconds } from "../../../../../../helpers/timeUtil";

const RebaseTimer: React.FC = () => {
  const { data: nextRebaseDate } = useNextRebaseDate();

  return (
    <Box className="rebase-timer">
      <Typography variant="body2">
        {nextRebaseDate ? (
          <>
            <strong>{prettifySeconds((nextRebaseDate.getTime() - new Date().getTime()) / 1000)}&nbsp;</strong>
            <Trans>to next rebase</Trans>
          </>
        ) : (
          <Skeleton width="155px" />
        )}
      </Typography>
    </Box>
  );
};

export default RebaseTimer;
