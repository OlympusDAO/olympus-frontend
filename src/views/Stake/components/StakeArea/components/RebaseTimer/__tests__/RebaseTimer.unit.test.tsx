import React from "react";
import RebaseTimer from "src/views/Stake/components/StakeArea/components/RebaseTimer/RebaseTimer";
import { describe, expect, it } from "vitest";

describe("<RebaseTimer/>", () => {
  it("should render component", () => {
    const container = <RebaseTimer />;
    expect(container).toMatchSnapshot();
  });
});
