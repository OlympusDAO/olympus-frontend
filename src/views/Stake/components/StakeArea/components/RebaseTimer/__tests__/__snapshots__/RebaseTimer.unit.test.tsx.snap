// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RebaseTimer/> should render component 1`] = `
<div>
  <div
    class="rebase-timer MuiBox-root css-0"
  >
    <p
      class="MuiTypography-root MuiTypography-body2 css-112gc2l-MuiTypography-root"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1l7q9tc-MuiSkeleton-root"
        style="width: 155px;"
      />
    </p>
  </div>
</div>
`;
